util/src/matrix.rs

```diff
@@ -0,0 +1,527 @@
+    ///
+    /// This function allocates an uninitialized matrix of with a given `height` and `width`. A
+    /// reference to this uninitialized array is passed to the user-provided closure, which is
+    /// responsible for initializing all elements.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe as `f` must initialization on every element of the matrix.
+    pub unsafe fn new_with<F: FnOnce(&mut Matrix<MaybeUninit<T>>)>(
+        height: usize,
+        width: usize,
+        f: F,
+    ) -> Self {
+        let mut res = Self::new_uninit(height, width);
+        f(&mut res);
+        // SAFETY: it's up to the user to ensure that `f` correctly initializes `res`.
+        res.assume_init()
+    }
+
+    /// The height of the matrix.
+    pub fn height(&self) -> usize {
+        self.height
+    }
+
+    /// The width of the matrix.
+    pub fn width(&self) -> usize {
+        self.width
+    }
+
+    /// The total size of the matrix, in terms of instances of `T`.
+    pub fn size(&self) -> usize {
+        debug_assert_eq!(self.height * self.width, self.data.len());
+        self.data.len()
+    }
+
+    /// Get a reference to the matrix as a row-major slice.
+    pub fn as_flat(&self) -> &[T] {
+        &self.data
+    }
+
+    /// Get a mutable reference to the matrix as a row-major slice.
+    pub fn as_flat_mut(&mut self) -> &mut [T] {
+        &mut self.data
+    }
+
+    /// Get a reference to a row of the matrix as a slice, returning `None` if out of bounds.
+    pub fn get(&self, index: usize) -> Option<&[T]> {
+        self.as_flat()
+            .get(index * self.width..(index + 1) * self.width)
+    }
+
+    /// Get a mutable reference to a row of the matrix as a slice, returning `None` if out of
+    /// bounds.
+    pub fn get_mut(&mut self, index: usize) -> Option<&mut [T]> {
+        let width = self.width; // borrow checker is not very smart
+        self.as_flat_mut()
+            .get_mut(index * width..(index + 1) * width)
+    }
+
+    /// Get a reference to a row of the matrix as a slice, without bounds checks.
+    ///
+    /// # Safety
+    ///
+    /// The `index` must be valid, i.e. `index < matrix.height()`.
+    pub unsafe fn get_unchecked(&self, index: usize) -> &[T] {
+        self.as_flat()
+            .get_unchecked(index * self.width..(index + 1) * self.width)
+    }
+
+    /// Get a mutable reference to a row of the matrix as a slice, without bounds checks.
+    ///
+    /// # Safety
+    ///
+    /// The `index` must be valid, i.e. `index < matrix.height()`.
+    pub unsafe fn get_unchecked_mut(&mut self, index: usize) -> &mut [T] {
+        let width = self.width; // borrow checker again
+        self.as_flat_mut()
+            .get_unchecked_mut(index * width..(index + 1) * width)
+    }
+
+    /// Iterate over all rows of the matrix as references.
+    pub fn iter(
+        &self,
+    ) -> impl Iterator<Item = &[T]> + DoubleEndedIterator + ExactSizeIterator + FusedIterator {
```

Comment:
Seems easiest to return `Chunks` (or `ChunksMut` below)

Reply:
We do return `Chunks` but that's not something I'd like the API to _promise_. I'd rather keep the flexibility to define our own iterator struct if we need extra functionality.